Enemy

gaebos opened the pull request.

gaebos committed:
Refactored color palette loading a bit

gaebos committed:
* Modified build script to just compile the main.asm for correct usage of import files
* First enemy animation -> still hardcoded animation

gaebos committed:
First version of an enemy set with configurable animations

gaebos committed:
Refactored enemy setup

gaebos committed:
Updated rom config

gaebos committed:
Fixed timing issue with anims

gaebos committed:
Refactored color palette loading to use params

gaebos committed:
loading different enemies of different tile size

gaebos committed:
Adding math macros

gaebos committed:
Added enemy object animations

gaebos committed:
Adding comments, first step of cleaning

voidexp commented on src/video.asm:
```diff
@@ -1,7 +1,3 @@
-.include "nes.asm"
```
let's revert this back

gaebos replied:
the whole file I guess?

voidexp commented on src/enemy.asm:
```diff
@@ -0,0 +1,442 @@
+;
+; Animation settings
+;
+.code
```
why this goes into `.code`? seems like `.rodata` is a better place for this stuff

voidexp commented on src/enemy.asm:
```diff
@@ -0,0 +1,442 @@
+
+.segment "RAM"
+; 6 enemies, 5 bytes each
+current_enemy_set:  .res 40 ; 2 byte enemy address, 2 byte address, 1 byte anim state
```
but 6x5 makes 30... :/

gaebos replied:
I have an excellent explanation for that!

voidexp commented on src/enemy.asm:
```diff
@@ -0,0 +1,442 @@
+;
```
When it comes to use of global variables such as temp_1 and temp_address, all this code becomes NMI-unsafe.
In interrupt can occur after any instruction and if the interrupt handler itself modifies anything in temp_-stuff, after the execution returns to the subroutine, the data will be in unknown state and everything will be screwed.

The code should be rewritten to be context independent and use only locals on stack. This should be discussed and a generic mechanism developed.

gaebos committed:
Removed virtual stack

gaebos committed:
Adding animation file

gaebos committed:
Adding testbed for testing macros. simple code etc

gaebos committed:
ECS with Projectiles created as entity with a sprite and movement component

gaebos committed:
Converted enemies to entities

gaebos committed:
* Adding collision detection for enemies
* Adding enemy component to increase kill_count
* Showing kill count

gaebos committed:
* position overflow detection

gaebos committed:
* Upgraded set of Spacetopuses
* Improved documentation

gaebos committed:
* Kill count working again

gaebos committed:
Updated assets for enemies

gaebos committed:
Merge branch 'enemy' of https://github.com/V0idExp/ShootEmUp into enemy

gaebos committed:
* proper disappearing of enemies on collision
* kill count fix
* changed background to one scrolling texture

gaebos committed:
* Last polish of kill_count

gaebos committed:
Removed outdated file and subroutines

gaebos committed:
Replaced usages of RAM with BSS

gaeb committed:
Adding of the
* player entity
* actor component
* rainbow entity
* flame entity

gaeb committed:
Don't update movement component if disabled
Fill all empty sprites with empty tile id after processing all sprite components

voidexp committed:
Use sys.interpreter for tools

voidexp committed:
Merge branch 'master' into enemy
